[DOCS] added sending events section in docs (#9064)
meshery/meshery: docs/pages/guides/send-event-in-server.md

---
layout: default
title: Send Events From Meshery Server Using Golang
description: This guide is to help backend contributors how to send server events using Golang.
permalink: guides/send-events
type: Guides
language: en
---

> &emsp;&emsp;&emsp;Meshery incorporates an internal events publication mechanism that provides users with real-time updates on the processes occurring within the Meshery server when interacting with its endpoints. It ensures that users are kept in the loop regarding the ongoing activities within the API, and guides users towards future steps to resolve issues. This guide will provide step-by-step instructions on sending events from the server, including when to trigger events and what information to include.

Reviewer: ```suggestion
Meshery incorporates an internal events publication mechanism that provides users with real-time updates on the processes occurring within the Meshery server when interacting with its endpoints. It ensures that users are kept in the loop regarding the ongoing activities within the API, and guides users towards future steps to resolve issues. This guide will provide step-by-step instructions on sending events from the server, including when to trigger events and what information to include.
```

> &emsp;&emsp;&emsp;Meshery incorporates an internal events publication mechanism that serves to provide users with real-time updates on the processes occuring within the Meshery API when they interact with its endpoints. it ensures that users are kept in the loop regarding the ongoing activities within the API, and guides users towards future steps to resolve issues. This guide will provide step-by-step instructions on how to send events from the server codebase, including when to trigger events and what information to include in.

Reviewer: ```suggestion
Meshery incorporates an internal events publication mechanism that serves to provide users with real-time updates on the processes occuring within the Meshery API when they interact with its endpoints. it ensures that users are kept in the loop regarding the ongoing activities within the API, and guides users towards future steps to resolve issues. This guide will provide step-by-step instructions on how to send events from the server codebase, including when to trigger events and what information to include in.
```

> &emsp;&emsp;&emsp;Meshery incorporates an internal events publication mechanism that serves to provide users with real-time updates on the processes occuring within the Meshery API when they interact with its endpoints. it ensures that users are kept in the loop regarding the ongoing activities within the API, and guides users towards future steps to resolve issues. This guide will provide step-by-step instructions on how to send events from the server codebase, including when to trigger events and what information to include in.

Reviewer 2: ``` suggestion
&emsp;&emsp;&emsp;Meshery incorporates an internal events publication mechanism that provides users with real-time updates on the processes occurring within the Meshery server when interacting with its endpoints. It ensures that users are kept in the loop regarding the ongoing activities within the API, and guides users towards future steps to resolve issues. This guide will provide step-by-step instructions on sending events from the server, including when to trigger events and what information to include.
```

First, let's take a look at how the event object is constructed,

<pre class="codeblock-pre">
 <div class="codeblock">
 eventBuilder := events.NewEvent()
                  .ActedUpon(UUID)
                  .FromUser(UUID)
                  .FromSystem(UUID)
                  .WithCategory(string)
                  .WithAction(string)
                  .WithSeverity(EventSeverity)
                  .WithDescription(string)
                  .WithMetadata(map[string]interface{})
                  .Build()
  </div>
 </pre>
&nbsp;&nbsp;

> *Note: events is a package **(github.com/layer5io/meshkit/models/events)** from meshkit containing source code of event related functionality*

Reviewer: ```suggestion
*Note: `events` is a package [github.com/meshery/meshkit/models/events](https://github.com/meshery/meshkit) from MeshKit containing source code of event related functionality*
```

## An Example in Code

<pre class="codeblock-pre">
 <div class="codeblock">
func (h *Handler) KubernetesPingHandler(w http.ResponseWriter, req *http.Request, _ *models.Preference, user *models.User, provider models.Provider) {
	userID := uuid.FromStringOrNil(user.ID)
	token, ok := req.Context().Value(models.TokenCtxKey).(string)
	if !ok {
		w.WriteHeader(http.StatusInternalServerError)
		fmt.Fprintf(w, "failed to get the token for the user")
		return
	}

	connectionID := req.URL.Query().Get("connection_id")
	eventBuilder := events.NewEvent().FromUser(userID).FromSystem(*h.SystemID).WithCategory("connection").WithAction("ping")
	if connectionID != "" {
		// Get the context associated with this ID
		k8sContext, err := provider.GetK8sContext(token, connectionID)
		eventBuilder.ActedUpon(uuid.FromStringOrNil(connectionID))
		if err != nil {
			eventBuilder.WithSeverity(events.Error).WithDescription(fmt.Sprintf("Encountered an issue while retrieving kubernetes context for connection ID `%s`", connectionID)).
				WithMetadata(map[string]interface{}{
					"error":                 err,

> 					"probable_cause":        "Meshery might not be able to access Kubeconfig.",

Reviewer 2: This is incorrect @zakisk, the severity of the event is an error, therefore the error should be MeshKit error, and the error would contain `probable_cause` and `suggested_remediation`.

Please update the example. Or add a snippet/link from the existing codebase.
When you add a link make sure to not source from the master branch but instead from one of the release tags. As in future, there will be changes in the code, which would result in invalid reference.

If `provider.GetK8sContext` succeeds, an event with a [Success](https://github.com/meshery/meshkit/blob/ea3c60907a1cd1902902a4113206579992772083/models/events/events.go#L21) severity level is constructed. The event includes a success message and metadata containing the context name and the address of the Kubernetes API server, obtained from the operation. This information aims to provide users with relevant details about the successful operation.

after constructing the Event object using the Build function, event is stored into database for records and subsequently published in a new goroutine through [EventBroadcaster](https://github.com/meshery/meshery/blob/995ab671a12013088f874430cfa2c0f025b073d2/server/models/handlers.go#L241) member variable of [HandlerConfig](https://github.com/meshery/meshery/blob/995ab671a12013088f874430cfa2c0f025b073d2/server/models/handlers.go#L206), which is part of [Handler](https://github.com/meshery/meshery/blob/995ab671a12013088f874430cfa2c0f025b073d2/server/handlers/handler_instance.go#L18) instance.

> 

Reviewer: @ptdatta have you documented the Notification Formatters, yet? A link and description of them needs to be included here.

Reviewer 3: No Currently I am working on the meshmodel UI. OK I'll do it by today.

Reviewer 3: @leecalcote  Done✅ [#9086](https://github.com/meshery/meshery/pull/9086)

Reviewer 4: Thanks, @ptdatta
